### PR TITLE
[DMS-259] reference validation response

### PR DESCRIPTION
**Pull request by bradbanister**

### CODE REVIEW

File: src/core/EdFi.DataManagementService.Core/Middleware/ParseBodyMiddleware.cs [hunk 1]
```diff
@@ -4,16 +4,38 @@
...
+using static EdFi.DataManagementService.Core.Response.FailureResponse;
 
 namespace EdFi.DataManagementService.Core.Middleware
 {
     internal class ParseBodyMiddleware(ILogger _logger) : IPipelineStep
     {
+        private static readonly JsonSerializerOptions _serializerOptions =
+            new() { Encoder = JavaScriptEncoder.UnsafeRelaxedJsonEscaping };
+
+        public static string GenerateFrontendErrorResponse(string errorDetail)
```
Comment (bradbanister) [inline src/core/EdFi.DataManagementService.Core/Middleware/ParseBodyMiddleware.cs:23]: moved to the only place it is used

File: src/core/EdFi.DataManagementService.Core/Response/FailureResponseWithErrors.cs [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    /// </summary>
+    int status,
+    /// <summary>
+    /// Gets or sets a correlation id that is logged with the details of the exception.
+    /// </summary>
+    string? correlationId,
+    /// <summary>
+    /// Contains arrays of validation errors keyed by JSON Path (intended for use by a domain expert).
+    /// </summary>
+    Dictionary<string, string[]>? validationErrors,
```
Comment (bradbanister) [inline src/core/EdFi.DataManagementService.Core/Response/FailureResponseWithErrors.cs:42]: I don't know if `validationErrors` and `errors` are part of the response anymore as of ODS/API 7.2, but they are not in the responses for invalid references.

File: src/tests/EdFi.DataManagementService.Tests.E2E/Extensions/DataTableExtensions.cs [hunk 3]
```diff
@@ -1,49 +0,0 @@
...
-// See the LICENSE and NOTICES files in the project root for more information.
-
-using System.Text.RegularExpressions;
-using Reqnroll;
-
-namespace EdFi.DataManagementService.Tests.E2E.Extensions;
-
-internal static class DataTableExtensions
-{
-    public static IEnumerable<Dictionary<string, object>> ExtractDescriptors(this DataTable dataTable)
```
Comment (bradbanister) [inline src/tests/EdFi.DataManagementService.Tests.E2E/Extensions/DataTableExtensions.cs:13]: Moved to StepDefinitions and simplified

File: src/tests/EdFi.DataManagementService.Tests.E2E/Features/References/InvalidReferenceValidation.feature.disabled [hunk 4]
```diff
@@ -1,154 +0,0 @@
-Feature: Invalid Reference Validation
```
Comment (bradbanister) [inline src/tests/EdFi.DataManagementService.Tests.E2E/Features/References/InvalidReferenceValidation.feature.disabled:1]: renamed

File: src/tests/EdFi.DataManagementService.Tests.E2E/Features/Resources/DeleteResourcesValidation.feature [hunk 5]
```diff
@@ -42,21 +42,21 @@ Feature: Resources "Delete" Operation validations
 
 
         Scenario: 05 Verify response when deleting a referenced school
-             Given the system has these "schools" references
+            Given the system has these "schools" references
```
Comment (bradbanister) [inline src/tests/EdFi.DataManagementService.Tests.E2E/Features/Resources/DeleteResourcesValidation.feature:45]: No idea why my editor did this, but some setting is feature file aware and auto-aligns and inserts bars

File: src/tests/EdFi.DataManagementService.Tests.E2E/Hooks/SetupHooks.cs [hunk 6]
```diff
@@ -46,6 +46,11 @@ public static async Task BeforeFeature(PlaywrightContext context, TestLogger log
     [AfterFeature]
     public static async Task AfterFeature(PlaywrightContext context, TestLogger logger)
     {
+        if (ContainerSetup.ApiContainer == null || ContainerSetup.DbContainer == null)
+        {
+            return;
+        }
+
```
Comment (bradbanister) [inline src/tests/EdFi.DataManagementService.Tests.E2E/Hooks/SetupHooks.cs:53]: Fixes out-of-container E2E crashes

File: src/tests/EdFi.DataManagementService.Tests.E2E/StepDefinitions/StepDefinitions.cs [hunk 7]
```diff
@@ -56,20 +56,34 @@ public void GivenThereAreNoSchools()
             //throw new PendingStepException();
         }
 
-        private async Task<List<IAPIResponse>> ProcessData(string entityType, DataTable dataTable)
+        private static (string, Dictionary<string, object>) ExtractDescriptorBody(string descriptorValue)
         {
-            var _apiResponses = new List<IAPIResponse>();
-            var baseUrl = $"data/ed-fi";
+            // build the descriptor object with string splitting operations
```
Comment (bradbanister) [inline src/tests/EdFi.DataManagementService.Tests.E2E/StepDefinitions/StepDefinitions.cs:61]: This is the core of the descriptor stuff moved from the extension file

File: src/tests/EdFi.DataManagementService.Tests.E2E/StepDefinitions/StepDefinitions.cs [hunk 8]
```diff
@@ -89,22 +103,50 @@ private async Task<List<IAPIResponse>> ProcessData(string entityType, DataTable
...
-            var _apiResponses = await ProcessData(entityType, dataTable);
+            var _apiResponses = await ProcessDataTable(entityType, dataTable);
+
+            _logger.log.Information($"Responses for Given(the system has these {entityType})");
 
             // Verify all the responses
             foreach (var response in _apiResponses)
             {
-                response.Status.Should().BeOneOf([201, 200]);
+                _logger.log.Information(response.TextAsync().Result);
```
Comment (bradbanister) [inline src/tests/EdFi.DataManagementService.Tests.E2E/StepDefinitions/StepDefinitions.cs:113]: Added a lot of logging for debug purposes

File: src/tests/EdFi.DataManagementService.Tests.E2E/StepDefinitions/StepDefinitions.cs [hunk 9]
```diff
@@ -89,22 +103,50 @@ private async Task<List<IAPIResponse>> ProcessData(string entityType, DataTable
...
+            }
+        }
+
+        [Given("the system has these descriptors")]
+        public async Task GivenTheSystemHasTheseDescriptors(DataTable dataTable)
+        {
+            _logger.log.Information($"Responses for Given(the system has these descriptors)");
```
Comment (bradbanister) [inline src/tests/EdFi.DataManagementService.Tests.E2E/StepDefinitions/StepDefinitions.cs:120]: Gave descriptor inserts explicit step definition

**Review by simpat-adam — APPROVED** [review]

**Pull request merged**